File: apps/mobile/src/lib/feedRanking.ts
/** Rank main-feed posts using engagement metadata, recency decay, and user interests. */

import { getPostTopics, getTopicLabel, ContentTopic } from './contentTopics';
import { InterestScores } from './userInterests';

export interface RankablePost {
  like_count?: number;
  comment_count?: number;
  view_count?: number;
  created_at?: string;
  category?: string | null;
  caption?: string | null;
  topics?: string[] | null;
  media_type?: string | null;
}

export function engagementScore(post: RankablePost): number {
  const likes = post.like_count ?? 0;
  const comments = post.comment_count ?? 0;
  const views = post.view_count ?? 0;
  const createdAt = post.created_at ? new Date(post.created_at).getTime() : Date.now();
  const ageHours = Math.max(0, (Date.now() - createdAt) / 3600000);
  const recencyDecay = 1 / (1 + ageHours / 48);
  return (likes * 3 + comments * 5 + views * 0.1) * recencyDecay;
}

export function interestBoost(post: RankablePost, scores: InterestScores): number {
  const topics = getPostTopics(post);
  let affinity = 0;
  for (const topic of topics) {
    affinity += scores[topic] ?? 0;
  }
  if (affinity <= 0) return 1;
  return 1 + Math.min(affinity * 0.018, 1.35);
}

export function personalizedScore(post: RankablePost, scores: InterestScores): number {
  return engagementScore(post) * interestBoost(post, scores);
}

export function rankPostsByEngagement<T extends RankablePost>(posts: T[]): T[] {
  return [...posts].sort((a, b) => engagementScore(b) - engagementScore(a));
}

export function rankPostsForUser<T extends RankablePost>(posts: T[], scores: InterestScores): T[] {
  if (Object.keys(scores).length === 0) return rankPostsByEngagement(posts);
  return [...posts].sort((a, b) => personalizedScore(b, scores) - personalizedScore(a, scores));
}

/** Best-matching topic for a post given current user affinities (for UI hints). */
export function getMatchingInterest(
  post: RankablePost,
  scores: InterestScores,
): { topic: string; label: string } | null {
  const topics = getPostTopics(post);
  let best: { topic: string; score: number } | null = null;
  for (const topic of topics) {
    const s = scores[topic] ?? 0;
    if (s > 1 && (!best || s > best.score)) best = { topic, score: s };
  }
  if (!best) return null;
  return { topic: best.topic, label: getTopicLabel(best.topic as ContentTopic) };
}

File: backend/app/services/feed_ranking.py
"""Rank feed posts using engagement metadata and recency decay."""

from __future__ import annotations

from datetime import datetime, timezone
from typing import Any


def _to_datetime(value: datetime | str | None) -> datetime:
    if isinstance(value, datetime):
        return value if value.tzinfo else value.replace(tzinfo=timezone.utc)
    if isinstance(value, str):
        return datetime.fromisoformat(value.replace("Z", "+00:00"))
    return datetime.now(timezone.utc)


def engagement_score(
    *,
    like_count: int = 0,
    comment_count: int = 0,
    view_count: int = 0,
    created_at: datetime | str | None = None,
) -> float:
    """Higher score = more relevant for the main feed."""
    created = _to_datetime(created_at)
    age_hours = max(0.0, (datetime.now(timezone.utc) - created).total_seconds() / 3600)
    recency_decay = 1.0 / (1.0 + age_hours / 48.0)
    engagement = like_count * 3 + comment_count * 5 + view_count * 0.1
    return engagement * recency_decay


def rank_posts(posts: list[dict[str, Any]]) -> list[dict[str, Any]]:
    return sorted(
        posts,
        key=lambda post: engagement_score(
            like_count=int(post.get("like_count") or 0),
            comment_count=int(post.get("comment_count") or 0),
            view_count=int(post.get("view_count") or 0),
            created_at=post.get("created_at"),
        ),
        reverse=True,
    )


def feed_rank_order_sql(table_alias: str = "p") -> str:
    alias = table_alias
    return f"""(
        ({alias}.like_count * 3 + {alias}.comment_count * 5 + COALESCE({alias}.view_count, 0) * 0.1)
        / (1 + EXTRACT(EPOCH FROM (NOW() - {alias}.created_at)) / 172800.0)
    ) DESC, {alias}.created_at DESC"""
